File: platforms/android/contribute/src/main/java/pulse/contribute/android/util/PromiseEvent.kt
package pulse.contribute.android.util

class PromiseEvent(private val promiseEvent: String){

    companion object Static {
        //gets defined internally on App Start
        var promiseEventResolver: ((Any?, Boolean, String) -> Unit)? = null
        var promiseEventDisposer: ((String) -> Unit)? = null
    }

    var isValid: Boolean = (promiseEvent != "[UNDEFINED_EVENT]")

    fun trigger(returnValue: Any?, isLastTrigger: Boolean = false) {
        if(promiseEvent == "[UNDEFINED_EVENT]"){ return }
        promiseEventResolver?.invoke(returnValue, isLastTrigger, promiseEvent)
    }

    fun dispose() {
        if(promiseEvent == "[UNDEFINED_EVENT]"){ return }
        promiseEventDisposer?.invoke(promiseEvent)
    }

}

File: platforms/dot-net/PromiseEvent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PulseContribute.DotNet
{
    public class PromiseEvent
    {
        private string promiseId;

        public static Action<object, bool, string>? _promiseEventResolver;
        public static Action<string>? _promiseEventDisposer;

        public PromiseEvent(string promiseRequest)
        {
            promiseId = promiseRequest;
        }

        public void Trigger(object returnValue, bool isLastTrigger = false)
        {
            if (promiseId == "[UNDEFINED_EVENT]") { return; }
            _promiseEventResolver?.Invoke(returnValue, isLastTrigger, promiseId);
        }

        public void Dispose()
        {
            if (promiseId == "[UNDEFINED_EVENT]") { return; }
            _promiseEventDisposer?.Invoke(promiseId);
        }

    }

}
